subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/NonHierarchicalFileWatcherUpdater.java

```diff
@@ -91,38 +90,32 @@ protected SnapshotHierarchy doUpdateVfsOnBuildStarted(SnapshotHierarchy root) {
 
     @Override
     public SnapshotHierarchy updateVfsOnBuildFinished(SnapshotHierarchy root, WatchMode watchMode, int maximumNumberOfWatchedHierarchies) {
-        WatchableHierarchies.Invalidator invalidator = (location, currentRoot) -> {
+        SnapshotHierarchy newRoot = super.updateVfsOnBuildFinished(root, watchMode, maximumNumberOfWatchedHierarchies);
+        LOGGER.info("Watching {} directories to track changes", watchedDirectories.entrySet().size());
+        return newRoot;
+    }
+
+    @Override
+    protected WatchableHierarchies.Invalidator createInvalidator() {
+        return (location, currentRoot) -> {
             SnapshotCollectingDiffListener diffListener = new SnapshotCollectingDiffListener();
             SnapshotHierarchy invalidatedRoot = currentRoot.invalidate(location, diffListener);
             diffListener.publishSnapshotDiff((removedSnapshots, addedSnapshots) -> virtualFileSystemContentsChanged(removedSnapshots, addedSnapshots, invalidatedRoot));
             return invalidatedRoot;
         };
-        SnapshotHierarchy newRoot = watchableHierarchies.removeUnwatchableContent(
-            root,
-            watchMode,
-            hierarchy -> containsSnapshots(hierarchy, root),
-            maximumNumberOfWatchedHierarchies,
-            invalidator
-        );
-        LOGGER.info("Watching {} directories to track changes", watchedDirectories.entrySet().size());
-        return newRoot;
     }
 
     @Override
-    public Collection<Path> getWatchedHierarchies() {
-        return watchableHierarchies.getWatchableHierarchies();
+    protected void startWatchingHierarchies(Collection<File> hierarchiesToWatch) {
```

Comment:
It looks like we should start watching the directory which contains the probe here, right? And probably stop watching the directory when we stop watching the hierarchy.

Reply:
You're right, added.